src/cargo/util/graph.rs

```diff
@@ -1,52 +1,284 @@
+    len_edges: usize,
+    /// The number of stored nodes, increased when `add` is called or
+    /// if `add_edge` or `link` is called with a previously unseen node.
+    len_nodes: usize,
+}
+
+impl<N: Clone, E: Clone> Graph<N, E> {
+    fn len(&self) -> GraphAge {
+        GraphAge {
+            len_edges: self.edges.len(),
+            len_nodes: self.nodes.len(),
+        }
+    }
 }
 
-impl<N: Eq + Ord + Clone, E: Default + Clone> Graph<N, E> {
+impl<N: Clone + Eq + Hash, E: Clone + PartialEq> Graph<N, E> {
     pub fn new() -> Graph<N, E> {
         Graph {
-            nodes: im_rc::OrdMap::new(),
+            edges: vec![],
+            nodes: Default::default(),
+        }
+    }
+
+    /// This resets a Graph to the same state as it was when the passed `age` was made.
+    /// All the other `&mut` methods are guaranteed to increase the `len` of the Graph.
+    /// So the reset can be accomplished by throwing out all newer items and fixing internal pointers.
+    fn reset_to(&mut self, age: GraphAge) {
+        // the prefix we are resetting to had `age.len_nodes`, so remove all newer nodes
+        assert!(self.nodes.len() >= age.len_nodes);
+        // IndexMap dose not have a `truncate` so we roll our own
+        while self.nodes.len() > age.len_nodes {
+            self.nodes.pop();
+        }
+
+        // the prefix we are resetting to had `age.len_edges`, so remove all links pointing to newer edges
+        for (_, lookup) in self.nodes.iter_mut() {
+            // IndexMap dose not have a `last` so we roll our own
+            while lookup.len() >= 1
+                && lookup
+                    .get_index(lookup.len() - 1)
+                    .filter(|(_, idx)| idx >= &&age.len_edges)
+                    .is_some()
+            {
+                lookup.pop();
+            }
+        }
+
+        // the prefix we are resetting to had `age.len_edges`, so
+        assert!(self.edges.len() >= age.len_edges);
+        // remove all newer edges and record the references that need to be fixed
+        let to_fix: Vec<EdgeIndex> = self
+            .edges
+            .drain(age.len_edges..)
+            .filter_map(|e| e.previous)
+            .filter(|idx| idx < &age.len_edges)
+            .collect();
+
+        // fix references into the newer edges we remove
+        for idx in to_fix {
+            self.edges[idx].next = None;
         }
+        assert_eq!(self.len(), age);
     }
 
     pub fn add(&mut self, node: N) {
-        self.nodes.entry(node).or_insert_with(im_rc::OrdMap::new);
+        // IndexMap happens to do exactly what we need to keep the ordering correct.
+        // This can be undone as it will increase the `len_nodes` if `node` is new.
+        self.nodes.entry(node).or_insert_with(IndexMap::new);
     }
 
-    pub fn link(&mut self, node: N, child: N) -> &mut E {
-        self.nodes
+    /// connect `node`to `child` with out associating any data.
+    pub fn link(&mut self, node: N, child: N) {
```

Comment:
Not clear to me if link means create an edge or something else. I think in general I'm just a bit confused between links and edges